source/rust_verify_test/tests/hash.rs

```diff
@@ -0,0 +1,150 @@
+            v1.push(1u8); v1.push(2u8);
+            v2.push(3u8); v2.push(4u8);
+            h1.write(v1.as_slice()); h1.write(v2.as_slice());
+            h2.write(v1.as_slice()); h2.write(v2.as_slice());
+            assert(h1@ == seq![seq![1u8, 2u8], seq![3u8, 4u8]]);
+            let digest1 = h1.finish();
+            let digest2 = h2.finish();
+            assert(digest1 == digest2);
+        }
+    } => Ok(())
+}
+
+test_verify_one_file! {
+    #[test] test_hash_map verus_code! {
+        use std::collections::HashMap;
+        use vstd::prelude::*;
+        fn test()
+        {
+            broadcast use vstd::std_specs::hash::group_hash_axioms;
+            let mut m = HashMap::<u32, i8>::new();
+            assert(m@ == Map::<u32, i8>::empty());
+
+            m.insert(3, 4);
+            m.insert(6, -8);
+            assert(m@[3] == 4);
+
+            let b = m.contains_key(&3);
+            assert(b);
+
+            let n = m.len();
+            assert(n == 2);
+
+            let v = m.get(&6);
+            match v {
+                Some(v) => assert(*v == -8),
+                None => assert(false),
+            };
+
+            m.clear();
+            assert(!m@.contains_key(3));
+            let b = m.contains_key(&3);
+            assert(!b);
+        }
+    } => Ok(())
+}
+
+test_verify_one_file! {
+    #[test] test_hash_map_box verus_code! {
+        use std::collections::HashMap;
+        use vstd::prelude::*;
+
+        fn test()
+        {
+            broadcast use vstd::std_specs::hash::group_hash_axioms;
+            let mut m = HashMap::<Box<u32>, i8>::new();
+            assert(m@ == Map::<Box<u32>, i8>::empty());
+
+            let three = Box::<u32>::new(3);
+            let six = Box::<u32>::new(6);
+            m.insert(three, 4);
+            m.insert(six, -8);
+
+            let b = m.contains_key(&3);
+            assert(b);
+
+            let v = m.get(&6);
+            match v {
+                Some(v) => assert(*v == -8),
+                None => assert(false),
+            };
+
+            m.clear();
+            let b = m.contains_key(&3);
+            assert(!b);
+        }
+    } => Ok(())
+}
+
+test_verify_one_file! {
+    #[test] test_hash_map_struct verus_code! {
+        use core::hash::{Hash, Hasher};
+        use std::collections::HashMap;
+        use vstd::prelude::*;
+
+        #[derive(PartialEq, Eq)]
+        struct MyStruct
+        {
+            pub i: u16,
+            pub j: i32,
+        }
+
+        impl Hash for MyStruct
+        {
+            #[verifier::external_body]
+            fn hash<H>(&self, state: &mut H)
+                where
+                    H: Hasher
+            {
+                self.i.hash(state);
+                self.j.hash(state);
+            }
+        }
+
+        fn test()
+        {
+            broadcast use vstd::std_specs::hash::group_hash_axioms;
+            assume(vstd::std_specs::hash::does_type_have_deterministic_hash::<MyStruct>());
```

Comment:
Would it be worthwhile to add a negative test, similar to this one, where you don't include the assumption about `MyStruct`?

Reply:
Sure. Though hopefully eventually we won't need such an assume, at which point we can remove the test.